src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -304,3 +284,63 @@ func didInstallationFinish(clusterStatus string, log logrus.FieldLogger, status
 	}
 	return false
 }
+
+func getClusterAndDetermineTaintRemoval(ctx context.Context, assistedController assistedinstallercontroller.Controller,
+	ic inventory_client.InventoryClient, kc k8s_client.K8SClient, logger logrus.FieldLogger, invoker string) (*models.Cluster, bool) {
+	var cluster *models.Cluster
+	removeUninitializedTaint := false
+	if invoker == common.InvokerAgent {
+		if Options.ControllerConfig.HighAvailabilityMode == models.ClusterHighAvailabilityModeNone {
+			// When the agent-based installer installs a SNO cluster, assisted-service
+			// will never be reachable because it will not be running after the boostrap
+			// node reboots to become the SNO cluster. SetReadyState will never be
+			// able to get a connection to assisted-service in this circumstance.
+			//
+			// Instead of exiting with panic with "invalid memory address or nil pointer
+			// dereference" and having the controller restart because cluster is nil,
+			// log warning and exit 0. Otherwise the controller will keep restarting
+			// leaving the controller in a running state, even through the cluster
+			// has finished install.
+			logger.Warnf("cluster is SNO and invoker = %v, skipping assisted-installer-controller", invoker)
+			os.Exit(0)
+		}
+		// With the agent-based installer, assisted-service runs on the bootstrap node.
+		// We do not need to wait for an extended amount of time to reach assisted-service
+		// because it should be running on the local network.
+		cluster = assistedController.SetReadyState(5 * time.Minute)
+		var platformType models.PlatformType
+		if cluster == nil {
+			// assisted-service may no longer be available after bootstrap node reboots
+			// and assisted-installer-controller is restarted. In this case, we try
+			// to determine the platform type by looking at the install-config stored
+			// in the cluster-config-v1 ConfigMap.
+			logger.Warnf("SetReadyState timed out fetching cluster from assisted-service, invoker = %v", invoker)
+			platformType, err := common.GetPlatformTypeFromInstallConfig(kc, logger)
+			if err != nil {
+				logger.Warnf("error determining platform type from install-config: %v", err)
+			} else {
+				logger.Infof("platform type from install-config: %v", platformType)
+			}
+		} else {
+			platformType = *cluster.Platform.Type
+		}
+
+		hasValidvSphereCredentials := false
+		if platformType == models.PlatformTypeVsphere {
+			hasValidvSphereCredentials := common.HasValidvSphereCredentials(ctx, ic, kc, logger)
+			if hasValidvSphereCredentials {
+				logger.Infof("Has valid vSphere credentials: %v", hasValidvSphereCredentials)
+			}
+		}
+		removeUninitializedTaint = common.RemoveUninitializedTaintAgent(platformType,
+			hasValidvSphereCredentials, Options.ControllerConfig.OpenshiftVersion)
+	} else {
+		cluster = assistedController.SetReadyState(assistedinstallercontroller.WaitTimeout)
```

Comment:
Should we just log in case cluster not found? I mean in not agent flow it is error that we should not continue no?

Reply:
I think today, in non agent flow, the controller fails and restarts again. But yes, let's log that cluster wasn't able to be retrieved to aid in debugging.